typed_sql/lib/src/codegen/parse_library.dart

```diff
@@ -0,0 +1,493 @@
+        element: a,
+      );
+    }
+
+    final field = ParsedField(
+      name: a.name,
+      documentation: a.documentationComment,
+      typeName: type,
+      isNullable: a.returnType.nullabilitySuffix != NullabilitySuffix.none,
+      backingType: backingType,
+      // TODO: Support Unique(given: [...])
+      unique: uniqueTypeChecker.hasAnnotationOf(a),
+      autoIncrement: autoIncrement,
+      defaultValue: _parseDefaultValue(a, type),
+    );
+    fields.add(field);
+  }
+
+  // Extract @References annotations
+  final foreignKeys = <ParsedForeignKey>[];
+  for (final a in cls.accessors) {
+    for (final annotation in referencesTypeChecker.annotationsOf(a)) {
+      final key = a.name;
+      final table = annotation.getField('table')?.toStringValue();
+      final field = annotation.getField('field')?.toStringValue();
+      final as = annotation.getField('as')?.toStringValue();
+      final name = annotation.getField('name')?.toStringValue();
+      if (table == null || field == null) {
+        throw InvalidGenerationSource(
+          'References annotation must have `table` and `field` fields!',
+          element: a,
+        );
+      }
+      if (fields.any((f) => f.name == name)) {
+        throw InvalidGenerationSource(
+          'References have `name: "$name"` which conflicts with the field '
+          'that has the same name!',
+          element: a,
+        );
+      }
+
+      final fk = ParsedForeignKey(
+        key: fields.firstWhere((f) => f.name == key),
+        table: table,
+        field: field,
+        as: as,
+        name: name,
+      );
+      foreignKeyToElement[fk] = a;
+      foreignKeys.add(fk);
+    }
+  }
+
+  // Extract primary key!
+  final pks = primaryKeyTypeChecker.annotationsOfExact(cls);
+  if (pks.isEmpty || pks.length > 1) {
+    throw InvalidGenerationSource(
+      'subclasses of `Model` must have one `PrimaryKey` annotation',
+      element: cls,
+    );
+  }
+  final primaryKey = pks.first
+      .getField('fields')!
+      .toListValue()!
+      .map((v) => v.toStringValue()!)
+      .map((key) {
+    final field = fields.firstWhereOrNull((field) => field.name == key);
+    if (field == null) {
+      throw InvalidGenerationSource(
+        'PrimaryKey annotation references unknown field "$field"',
+        element: cls,
+      );
+    }
+    return field;
+  }).toList();
+
+  if (primaryKey.isEmpty) {
+    throw InvalidGenerationSource(
+      'subclasses of `Model` must have a non-empty `PrimaryKey` annotation',
+      element: cls,
+    );
+  }
+  if (primaryKey.length != primaryKey.toSet().length) {
+    throw InvalidGenerationSource(
+      'subclasses of `Model` cannot have duplicate fields in `PrimaryKey`',
+      element: cls,
+    );
+  }
+  if (primaryKey.any((f) => f.isNullable)) {
+    throw InvalidGenerationSource(
+      'PrimaryKey fields cannot be nullable',
+      element: cls,
+    );
+  }
+
+  return ParsedModel(
+    name: cls.name,
```

Comment:
Do you want to validate the name? The main issue I can see is that dart identifiers can contain $, I don't think that is valid in sql (without quoting).

https://www.postgresql.org/docs/7.0/syntax525.htm

Same for field names.